File: src/SwapNodesInPairs.java
/**
 * User: guoxc
 * Date: 15-1-11
 * Time: 22:23
 */
public class SwapNodesInPairs {
    public ListNode swapPairs(ListNode head){
        ListNode res = head;
        ListNode current = head;
        ListNode old_c = head;
        if(current != null && current.next != null){
            ListNode c1 = current;
            ListNode c2 = current.next;
            ListNode c3 = current.next.next;

            res = c2;
            c2.next = c1;
            c1.next = c3;
            current = c3;
            old_c = c1;
        }
        while(current != null && current.next != null){
            ListNode c1 = current;
            ListNode c2 = current.next;
            ListNode c3 = current.next.next;

            old_c.next = c2;
            c2.next = c1;
            c1.next = c3;
            current = c3;
            old_c = c1;
        }
        return res;

    }
    public static void test(){
        ListNode l1 = new ListNode(new int[]{1,2,3,4});
        System.out.println(new SwapNodesInPairs().swapPairs(l1));
    }
}

File: cpp/swap_nodes_in_pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head == NULL || head->next == NULL) return head;
        ListNode* slow = head;
        ListNode* fast = head->next;
        ListNode* res = fast;
        ListNode* prev = slow;
        while(slow != NULL && fast != NULL){
            prev->next = fast;
            slow->next = fast->next;
            fast->next = slow;
            if(slow == NULL || slow->next == NULL) break;
            prev = slow;
            slow = slow->next;
            fast = slow->next;
        }
        return res;
    }
};
